Frequent Websocket and TCP Keep Alives keep LTE connection open and drain battery

Hi,

I have installed Gotify on an Ubuntu 18.04 server VM, no docker, no proxy, straight forward. The server works well but I notice that there is frequent TCP traffic, **around every 10 seconds** between the server and the Android client. This keeps the LTE connection to the network constantly in RRC-Connected state and has a significant impact on battery use.

I ran a Wireshark trace, see screenshot attach which shows that this is caused by two thing:

WebSocket keep alives (ping/pong)
TCP keepalives

Both keepalives are always triggered from the servers side.

Note: Other messaging systems such as Prosody/Conversations/XMPP only send tcp-keep alive messaging every 7-10 minutes. This is way more batter efficient.

Are there any settings to adapt Gotify's keep alive messaging frequency?

Thanks,
Martin

![2020-07-28_13-50-Gotify-tcp-ack and websocket ping](https://user-images.githubusercontent.com/47881430/88679695-f8e8f800-d0ef-11ea-999e-d29a7ebd43ca.png)

Ups, looks like the keepalive change breaks TLS connection establishment. Here's the output, seems to have happened in other Go projects before as well.

```
$ sudo ./gotify-linux-amd64
Starting Gotify version keepalive@2020-08-01-16:08:45
Started Listening for TLS connection on :7222
Started Listening for plain HTTP connection on :7221
panic: interface conversion: net.Conn is *tls.Conn, not *net.TCPConn

goroutine 11 [running]:
github.com/gotify/server/v2/runner.setKeepAlivePeriod.func1(0x1673620, 0xc00035aa80, 0x0)
        /proj/runner/runner.go:68 +0x10c
net/http.(*conn).setState(0xc0002ebb80, 0x1673620, 0xc00035aa80, 0x0)
        /usr/local/go/src/net/http/server.go:1725 +0xe4
net/http.(*Server).Serve(0xc000344380, 0x1667fe0, 0xc0003abc20, 0x0, 0x0)
        /usr/local/go/src/net/http/server.go:2932 +0x320
net/http.(*Server).ServeTLS(0xc000344380, 0x16690a0, 0xc0003aacc0, 0xc00026b3b0, 0x21, 0xc00026b3e0, 0x24, 0x0, 0x0)
        /usr/local/go/src/net/http/server.go:2973 +0x297
net/http.(*Server).ListenAndServeTLS(0xc000344380, 0xc00026b3b0, 0x21, 0xc00026b3e0, 0x24, 0x0, 0x0)
        /usr/local/go/src/net/http/server.go:3130 +0x11b
github.com/gotify/server/v2/runner.Run.func2(0xc000344380, 0xc0000ed980)
        /proj/runner/runner.go:44 +0x58
created by github.com/gotify/server/v2/runner.Run
        /proj/runner/runner.go:43 +0x2c7
```

Startup of the standard binary works o.k.:
```

sudo ./gotify-linux-amd64.original 
Starting Gotify version 2.0.17@2020-07-18-19:24:04
Started Listening for TLS connection on :7222
Started Listening for plain HTTP connection on :7221
[GIN] 2020/08/06 - 08:17:03 | 200 |    1.712239ms |   79.216.53.160 | GET      "/current/user"
[GIN] 2020/08/06 - 08:17:03 | 200 |     2.21981ms |   79.216.53.160 | GET      "/stream?token=xxxxxx"
[GIN] 2020/08/06 - 08:17:03 | 200 |    3.126486ms |   79.216.53.160 | GET      "/application"
[GIN] 2020/08/06 - 08:17:03 | 200 |    5.126883ms |   79.216.53.160 | GET      "/message?since=0"
[GIN] 2020/08/06 - 08:17:04 | 200 |    1.086982ms |   79.216.53.160 | GET      "/current/user"
```

A quick search revealed a similar problem in another project with a fix. Perhaps useful here as well:

https://github.com/etcd-io/etcd/issues/2184

--- a/api/stream/stream_test.go
+++ b/api/stream/stream_test.go
@@ -329,7 +329,7 @@ func TestMultipleClients(t *testing.T) {
 	userIDs := []uint{1, 1, 1, 2, 2, 3}
 	i := 0
 	server, api := bootTestServer(func(context *gin.Context) {
-		auth.RegisterAuthentication(context, nil, userIDs[i], "t"+string(userIDs[i]))
+		auth.RegisterAuthentication(context, nil, userIDs[i], "t"+fmt.Sprint(userIDs[i]))
 		i++
 	})
 	defer server.Close()

--- a/config.example.yml
+++ b/config.example.yml
@@ -2,6 +2,7 @@
 # Save it to `config.yml` when edited
 
 server:
+  keepaliveperiodseconds: 0 # 0 = use system default; set the interval in which keepalive packages will be sent. Only change this value if you know what you are doing.
   listenaddr: "" # the address to bind on, leave empty to bind on all addresses
   port: 80 # the port the HTTP server will listen on
 

--- a/config.example.yml
+++ b/config.example.yml
@@ -35,6 +35,7 @@ server:
 #      - "Authorization"
 #      - "content-type"
   stream:
+    pingperiodseconds: 45 # the interval in which websocket pings will be sent. Only change this value if you know what you are doing.
     allowedorigins: # allowed origins for websocket connections (same origin is always allowed)
 #      - ".+.example.com"
 #      - "otherdomain.com"

--- a/config/config.go
+++ b/config/config.go
@@ -11,9 +11,11 @@ import (
 // Configuration is stuff that can be configured externally per env variables or config file (config.yml).
 type Configuration struct {
 	Server struct {
-		ListenAddr string `default:""`
-		Port       int    `default:"80"`
-		SSL        struct {
+		KeepAlivePeriodSeconds int
+		ListenAddr             string `default:""`
+		Port                   int    `default:"80"`
+
+		SSL struct {
 			Enabled         *bool  `default:"false"`
 			RedirectToHTTPS *bool  `default:"true"`
 			ListenAddr      string `default:""`

--- a/config/config.go
+++ b/config/config.go
@@ -31,7 +31,8 @@ type Configuration struct {
 		}
 		ResponseHeaders map[string]string
 		Stream          struct {
-			AllowedOrigins []string
+			PingPeriodSeconds int `default:"45"`
+			AllowedOrigins    []string
 		}
 		Cors struct {
 			AllowOrigins []string

--- a/router/router.go
+++ b/router/router.go
@@ -25,7 +25,7 @@ func Create(db *database.GormDatabase, vInfo *model.VersionInfo, conf *config.Co
 	g.Use(gin.Logger(), gin.Recovery(), error.Handler(), location.Default())
 	g.NoRoute(error.NotFound())
 
-	streamHandler := stream.New(45*time.Second, 15*time.Second, conf.Server.Stream.AllowedOrigins)
+	streamHandler := stream.New(time.Duration(conf.Server.Stream.PingPeriodSeconds)*time.Second, 15*time.Second, conf.Server.Stream.AllowedOrigins)
 	authentication := auth.Auth{DB: db}
 	messageHandler := api.MessageAPI{Notifier: streamHandler, DB: db}
 	healthHandler := api.HealthAPI{DB: db}

--- a/runner/runner.go
+++ b/runner/runner.go
@@ -1,11 +1,14 @@
 package runner
 
 import (
+	"context"
 	"crypto/tls"
 	"fmt"
 	"log"
 	"net"
 	"net/http"
+	"strconv"
+	"time"
 
 	"github.com/gin-gonic/gin"
 	"github.com/gotify/server/v2/config"
@@ -18,7 +21,7 @@ func Run(engine *gin.Engine, conf *config.Configuration) {
 
 	if *conf.Server.SSL.Enabled {
 		if *conf.Server.SSL.RedirectToHTTPS {
-			httpHandler = redirectToHTTPS(string(conf.Server.SSL.Port))
+			httpHandler = redirectToHTTPS(strconv.Itoa(conf.Server.SSL.Port))
 		}
 
 		addr := fmt.Sprintf("%s:%d", conf.Server.SSL.ListenAddr, conf.Server.SSL.Port)
@@ -38,12 +41,24 @@ func Run(engine *gin.Engine, conf *config.Configuration) {
 		}
 		fmt.Println("Started Listening for TLS connection on " + addr)
 		go func() {
-			log.Fatal(s.ListenAndServeTLS(conf.Server.SSL.CertFile, conf.Server.SSL.CertKey))
+			listener := startListening(addr, conf.Server.KeepAlivePeriodSeconds)
+			log.Fatal(s.ServeTLS(listener, conf.Server.SSL.CertFile, conf.Server.SSL.CertKey))
 		}()
 	}
 	addr := fmt.Sprintf("%s:%d", conf.Server.ListenAddr, conf.Server.Port)
 	fmt.Println("Started Listening for plain HTTP connection on " + addr)
-	log.Fatal(http.ListenAndServe(addr, httpHandler))
+	server := &http.Server{Addr: addr, Handler: httpHandler}
+
+	log.Fatal(server.Serve(startListening(addr, conf.Server.KeepAlivePeriodSeconds)))
+}
+
+func startListening(addr string, keepAlive int) net.Listener {
+	lc := net.ListenConfig{KeepAlive: time.Duration(keepAlive) * time.Second}
+	conn, err := lc.Listen(context.Background(), "tcp", addr)
+	if err != nil {
+		log.Fatalln("Could not listen on", addr, err)
+	}
+	return conn
 }
 
 func redirectToHTTPS(port string) http.HandlerFunc {

